File: Motoshift/lib/models/nota_fiscal_filtro.dart
/// Os filtros da lista de notas fiscais, como a tela os guarda e a API os
/// recebe.
///
/// Mesmo desenho do [ExtratoFiltro]: um objeto imutável que sabe virar query
/// string. O período é o da COMPETÊNCIA — a data do serviço —, não o da
/// emissão: é por competência que se procura uma nota.
class NotaFiscalFiltro {
  const NotaFiscalFiltro({
    this.papel,
    this.competenciaDe,
    this.competenciaAte,
    this.status,
    this.contraparteId,
    this.turnoId,
  });

  /// `prestador` (notas em que eu prestei) ou `tomador` (serviços que tomei).
  final String? papel;
  final DateTime? competenciaDe;
  final DateTime? competenciaAte;

  /// `emitida` ou `cancelada`.
  final String? status;
  final int? contraparteId;
  final int? turnoId;

  bool get vazio =>
      papel == null &&
      competenciaDe == null &&
      competenciaAte == null &&
      status == null &&
      contraparteId == null &&
      turnoId == null;

  NotaFiscalFiltro copyWith({
    String? papel,
    DateTime? competenciaDe,
    DateTime? competenciaAte,
    String? status,
    int? contraparteId,
    int? turnoId,
    bool limparPapel = false,
    bool limparPeriodo = false,
    bool limparStatus = false,
    bool limparContraparte = false,
  }) {
    return NotaFiscalFiltro(
      papel: limparPapel ? null : (papel ?? this.papel),
      competenciaDe: limparPeriodo ? null : (competenciaDe ?? this.competenciaDe),
      competenciaAte: limparPeriodo ? null : (competenciaAte ?? this.competenciaAte),
      status: limparStatus ? null : (status ?? this.status),
      contraparteId: limparContraparte ? null : (contraparteId ?? this.contraparteId),
      turnoId: turnoId ?? this.turnoId,
    );
  }

  /// Só o que está preenchido entra na URL: parâmetro vazio é parâmetro a
  /// mais para o backend validar.
  String toQuery() {
    String dia(DateTime d) =>
        '${d.year.toString().padLeft(4, '0')}-'
        '${d.month.toString().padLeft(2, '0')}-'
        '${d.day.toString().padLeft(2, '0')}';

    final partes = <String>[
      if (papel != null) 'papel=$papel',
      if (competenciaDe != null) 'competenciaDe=${dia(competenciaDe!)}',
      if (competenciaAte != null) 'competenciaAte=${dia(competenciaAte!)}',
      if (status != null) 'status=$status',
      if (contraparteId != null) 'contraparteId=$contraparteId',
      if (turnoId != null) 'turnoId=$turnoId',
    ];
    return partes.join('&');
  }
}

File: backend/src/main/java/com/motoshift/dto/NotaFiscalFiltro.java
package com.motoshift.dto;

import java.time.LocalDate;
import java.time.LocalDateTime;

/**
 * Os filtros da lista de notas fiscais, como chegam na query string.
 *
 * <p>Mesmo desenho do {@link ExtratoFiltro}: cada campo é um parâmetro da API,
 * e as datas chegam como dia — o fim do intervalo inclui o dia inteiro.
 *
 * <ul>
 *   <li>{@code papel}: {@code prestador} (notas em que eu prestei) ou
 *       {@code tomador} (notas de serviços que eu tomei);</li>
 *   <li>{@code competenciaDe}/{@code competenciaAte}: período da DATA DO
 *       SERVIÇO, não da emissão — é o que o fisco pergunta;</li>
 *   <li>{@code status}: {@code emitida} ou {@code cancelada};</li>
 *   <li>{@code contraparteId}: o outro lado da nota;</li>
 *   <li>{@code turnoId}.</li>
 * </ul>
 */
public class NotaFiscalFiltro {

    private String papel;
    private LocalDate competenciaDe;
    private LocalDate competenciaAte;
    private String status;
    private Long contraparteId;
    private Long turnoId;

    public String getPapel() { return papel; }
    public void setPapel(String papel) { this.papel = papel; }

    public LocalDateTime getCompetenciaDe() {
        return competenciaDe == null ? null : competenciaDe.atStartOfDay();
    }

    /** Começo do dia SEGUINTE: a consulta usa {@code <}, então o dia final entra inteiro. */
    public LocalDateTime getCompetenciaAte() {
        return competenciaAte == null ? null : competenciaAte.plusDays(1).atStartOfDay();
    }

    public LocalDate getCompetenciaDeBruta() { return competenciaDe; }
    public LocalDate getCompetenciaAteBruta() { return competenciaAte; }
    public void setCompetenciaDe(LocalDate d) { this.competenciaDe = d; }
    public void setCompetenciaAte(LocalDate d) { this.competenciaAte = d; }

    public String getStatus() { return status; }
    public void setStatus(String status) { this.status = status; }

    public Long getContraparteId() { return contraparteId; }
    public void setContraparteId(Long contraparteId) { this.contraparteId = contraparteId; }

    public Long getTurnoId() { return turnoId; }
    public void setTurnoId(Long turnoId) { this.turnoId = turnoId; }
}
